kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,304 @@
+use super::{CacheId, GateType};
+use crate::circuits::{
+    expr::{Column, Expr, FeatureFlag, Op2, Variable},
+    gate::CurrOrNext,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum FoldingColumn {
+    Witness(usize),
+    ///for the extra columns added by quadricization
+    #[allow(dead_code)]
+    WitnessExtended(usize),
+    Index(GateType),
+    Coefficient(usize),
+    ///basically X, to allow accesing the next row
+    Shift,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+///A simplified expression to be use in folding
+pub enum FoldingExp<C> {
+    Constant(C),
+    Cell(FoldingColumn),
+    Double(Box<FoldingExp<C>>),
+    Square(Box<FoldingExp<C>>),
+    Add(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Sub(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Mul(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    UnnormalizedLagrangeBasis(i32),
+    Cache(CacheId, Box<FoldingExp<C>>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Degree {
+    Zero,
+    One,
+    Two,
+}
+impl<C> FoldingExp<C> {
+    pub(super) fn degree(&self) -> Degree {
+        use Degree::*;
+        match self {
+            FoldingExp::Constant(_) => Zero,
+            FoldingExp::Cell(col) => match col {
+                FoldingColumn::Witness(_) | FoldingColumn::WitnessExtended(_) => One,
+                _ => Zero,
+            },
+            FoldingExp::Double(e) => e.degree(),
+            FoldingExp::Square(e) => &e.degree() * &e.degree(),
+            FoldingExp::Add(e1, e2) | FoldingExp::Sub(e1, e2) | FoldingExp::Mul(e1, e2) => {
+                e1.degree() + e2.degree()
```

Comment:
Why is the degree of the addition and subtraction the sum of both sides' instead of the maximum between them? I understand for multiplication, but for the others?

Reply:
good question.  maybe it should be `max(e1.degree(), e2.degree())` ?